### ISSUE TITLE
Typescript `connect()` fails about circular references

### ISSUE THREAD

**Issue by semoal** [rematch/rematch#893]
**Describe the bug**

When updating Typescript from 4.1.x to 4.2.x, I've observed that `connect()` method exported from `react-redux` has a type error.

<img width="1154" alt="Screenshot 2021-05-16 at 7 25 52 PM" src="https://user-images.githubusercontent.com/22656541/118406589-af689a00-b67c-11eb-8758-9f028bff1f57.png">

**Expected behavior**

To not fail

**Additional context**

The possibly solution is to just use `any` instead of using RematchDispatch<RM>, I think this change doesn't break anything.

<img width="789" alt="Screenshot 2021-05-16 at 7 29 04 PM" src="https://user-images.githubusercontent.com/22656541/118406669-0ec6aa00-b67d-11eb-81e1-175936bf5f20.png">

**Comment by tianzhich** [comment]
I'll pick this one and try to solve it this week

**Comment by tianzhich** [comment]
I found the error cause, but not sure if this is an issue with typescript. 

Our `RematchDispatch`: 

```ts
export type RematchDispatch<
	TModels extends Models<TModels> = Record<string, any>
> = ReduxDispatch & ExtractRematchDispatchersFromModels<TModels>
```

If I removed the hybrid part of `ReduxDispatch`, the error gone.

**Comment by tianzhich** [comment]
I tried to create a minimal reproduce but I failed. If some could help more appreciated.

**Comment by semoal** [comment]
You have it in the count-react-ts file a repro.

**Comment by tianzhich** [comment]
> You have it in the count-react-ts example a repro. Be sure to update the typescript version to 4.2.x

You are right, but I mean to create a typescript repro based on a minimum of our types, though I'm not sure if it is an issue with typescript. I searched circular reference error, but few related issues were found.

**Comment by tbaliukynas** [comment]
I noticed that using `RematchRootState<RootModel>` and `RematchDispatch<RootModel>` directly instead of `RootState` and `Dispatch` defined in `store.ts` seems to fix the issue (at least in the `count-react-ts` example):
```diff
diff --git a/examples/count-react-ts/src/App.tsx b/examples/count-react-ts/src/App.tsx
index fb0774a..5b1eaa6 100644
--- a/examples/count-react-ts/src/App.tsx
+++ b/examples/count-react-ts/src/App.tsx
@@ -1,13 +1,14 @@
+import { RematchDispatch, RematchRootState } from '@rematch/core'
 import * as React from 'react'
 import { connect } from 'react-redux'
-import { RootState, Dispatch } from './store'
+import { RootModel } from './models'
 
-const mapState = (state: RootState) => ({
+const mapState = (state: RematchRootState<RootModel>) => ({
 	dolphins: state.dolphins,
 	sharks: state.sharks,
 })
 
-const mapDispatch = (dispatch: Dispatch) => ({
+const mapDispatch = (dispatch: RematchDispatch<RootModel>) => ({
 	incrementDolphins: () => dispatch.dolphins.increment(1),
 	incrementDolphinsAsync: dispatch.dolphins.incrementAsync,
 	incrementSharks: () => dispatch.sharks.increment(1),

```

**Comment by tianzhich** [comment]
Thanks for the reminder. It's only related to `dispatch`, using `RematchDispatch<Dispatch>` instead of `Dispatch` fix the issue.

**Comment by tianzhich** [comment]
It is strange. We may need help from a more advanced TSer.

**Comment by semoal** [comment]
Close this, since will be fixed with 2.1.0 release
